### DISCUSSION TITLE
Successful WASM integration in Jupyter notebooks

### DISCUSSION THREAD

**Discussion by basnijholt** [hpcc-systems/hpcc-js-wasm#287] [Show and tell]
Hi! I wanted to share a successful implementation of @hpcc-js/wasm in a Jupyter widget environment: https://github.com/pipefunc/graphviz-anywidget. I found @GordonSmith's issue here https://github.com/pygraphviz/pygraphviz/issues/319#issue-787662143 so I thought it might be nice to share.

We faced an interesting challenge, getting WASM to work consistently across different Jupyter environments (Lab, Notebook, VS Code). Our solution was to embed the WASM binary in the JS bundle as base64.

This approach works reliably across all Jupyter environments without needing a separate file server. I thought this might be useful information for others trying to use @hpcc-js/wasm in similar environments.

Thank you for providing this powerful WASM implementation!

**Comment by GordonSmith** [comment]
@basnijholt thanks for sharing this information!

It does sound like you might be using an older version of this library as the later versions adopt a similar approach except it compresses the wasm file using zstd and then base91 encodes the result.

Also worth mentioning that the newer versions have also split out the various wasm libraries into their own packages, with "@hpcc-js/wasm" existing just for backward compatibility.

**Reply by basnijholt** [reply]
@GordonSmith you are right! I just switched to the latest version and that made it so much easier. It allowed me to remove all of my hacks 😃
